mobile/src/main/java/org/openhab/habdroid/core/connection/ConnectionFactory.java

```diff
@@ -296,19 +307,59 @@ private static Connection determineAvailableConnection(Context context,
 
     @VisibleForTesting
     public synchronized void updateConnections() {
+        mCloudConnection = null;
+        if (mCloudConnectionCheckHandle != null) {
+            mCloudConnectionCheckHandle.cancel();
+            mCloudConnectionCheckHandle = null;
+        }
+
         if (settings.getBoolean(Constants.PREFERENCE_DEMOMODE, false)) {
             mLocalConnection = mRemoteConnection = new DemoConnection(ctx, settings);
             updateAvailableConnection(mLocalConnection, null);
+            CloudMessagingHelper.onConnectionUpdated(ctx, null);
         } else {
             mLocalConnection = makeConnection(Connection.TYPE_LOCAL, Constants.PREFERENCE_LOCAL_URL,
                     Constants.PREFERENCE_LOCAL_USERNAME, Constants.PREFERENCE_LOCAL_PASSWORD);
             mRemoteConnection = makeConnection(Connection.TYPE_REMOTE, Constants.PREFERENCE_REMOTE_URL,
                     Constants.PREFERENCE_REMOTE_USERNAME, Constants.PREFERENCE_REMOTE_PASSWORD);
 
+            if (mRemoteConnection != null) {
+                final MyAsyncHttpClient client = mRemoteConnection.getAsyncHttpClient();
+                mCloudConnectionCheckHandle = client.get(CloudConnection.SETTINGS_ROUTE, new MyHttpClient.TextResponseHandler() {
+                    @Override
+                    public void onFailure(Call call, int statusCode, Headers headers, String responseBody, Throwable error) {
+                        Log.e(TAG, "Error loading notification settings: " + error.getMessage());
+                        mCloudConnectionCheckHandle = null;
+                        CloudMessagingHelper.onConnectionUpdated(ctx, null);
+                    }
+
+                    @Override
+                    public void onSuccess(Call call, int statusCode, Headers headers, String responseBody) {
+                        try {
+                            JSONObject json = new JSONObject(responseBody);
+                            mCloudConnection = new CloudConnection(ctx, settings, mRemoteConnection, json);
+                            for (UpdateListener l : mListeners) {
+                                l.onConnectionSetChanged();
+                            }
+                        } catch (JSONException e) {
+                            Log.d(TAG, "Unable to parse notification settings JSON", e);
+                        }
+                        mCloudConnectionCheckHandle = null;
+                        CloudMessagingHelper.onConnectionUpdated(ctx, mCloudConnection);
+                    }
+                });
+            } else {
+                CloudMessagingHelper.onConnectionUpdated(ctx, null);
+            }
```

Comment:
We could also move all of this to the update handler thread and do the HTTP call synchronously there, so that available and cloud connection finally are updated together. Thoughts?

Reply:
Thinking about it I like this solution. Additionally we can add a blockingWaitForInitialization() call that the GCM service then can call; we then can also convert the voice service back to an IntentService and make it use that method.